### ISSUE TITLE
VALUES clause with CAST(ROW(...) AS ROW(...))` fails analysis

### ISSUE THREAD

**Issue by haozhun** [prestodb/presto#8255]
This query will fail analysis. It should not. It should produce a single row with one column of type `bigint`. (Note: The type is not `row(x bigint)` as one might think.)

```
select * from (
  values 
  (
    cast(row(1) as row(x bigint))
  )
) t(a);
```

Error message and stacktrace (with my annotation of where the problem is):

```
Query 20170613_165944_20212_jb4h6 failed: Cannot cast row(x bigint) to bigint
com.facebook.presto.sql.analyzer.SemanticException: Cannot cast row(x bigint) to bigint
        at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.visitCast(ExpressionAnalyzer.java:928)
        at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.visitCast(ExpressionAnalyzer.java:282)
        at com.facebook.presto.sql.tree.Cast.accept(Cast.java:98)
        at com.facebook.presto.sql.tree.StackableAstVisitor.process(StackableAstVisitor.java:26)
        at com.facebook.presto.sql.analyzer.ExpressionAnalyzer$Visitor.process(ExpressionAnalyzer.java:301)
        at com.facebook.presto.sql.analyzer.ExpressionAnalyzer.analyze(ExpressionAnalyzer.java:258)
        at com.facebook.presto.sql.planner.ExpressionInterpreter.evaluateConstantExpression(ExpressionInterpreter.java:219)
        at com.facebook.presto.sql.planner.RelationPlanner.visitValues(RelationPlanner.java:467)     <----- interesting, wrong branch
        at com.facebook.presto.sql.planner.RelationPlanner.visitValues(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.Values.accept(Values.java:55)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.QueryPlanner.planQueryBody(QueryPlanner.java:255)
        at com.facebook.presto.sql.planner.QueryPlanner.plan(QueryPlanner.java:130)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:433)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.RelationPlanner.visitTableSubquery(RelationPlanner.java:426)
        at com.facebook.presto.sql.planner.RelationPlanner.visitTableSubquery(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.TableSubquery.accept(TableSubquery.java:53)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.RelationPlanner.visitAliasedRelation(RelationPlanner.java:165)
        at com.facebook.presto.sql.planner.RelationPlanner.visitAliasedRelation(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.AliasedRelation.accept(AliasedRelation.java:71)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.QueryPlanner.planFrom(QueryPlanner.java:266)
        at com.facebook.presto.sql.planner.QueryPlanner.plan(QueryPlanner.java:150)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuerySpecification(RelationPlanner.java:440)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuerySpecification(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.QueryPlanner.planQueryBody(QueryPlanner.java:255)
        at com.facebook.presto.sql.planner.QueryPlanner.plan(QueryPlanner.java:130)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:433)
        at com.facebook.presto.sql.planner.RelationPlanner.visitQuery(RelationPlanner.java:95)
        at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.planner.LogicalPlanner.createRelationPlan(LogicalPlanner.java:368)
        at com.facebook.presto.sql.planner.LogicalPlanner.planStatementWithoutOutput(LogicalPlanner.java:172)
        at com.facebook.presto.sql.planner.LogicalPlanner.planStatement(LogicalPlanner.java:153)
        at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:126)
        at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:121)
        at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:312)
        at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:291)
        at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:247)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)
```

You'll notice in the function marked above that an improper coercion is somehow added to the analysis. This is where the incorrect annotation was introduced (with my annotation):

```
java.lang.Thread.State: RUNNABLE
        at com.facebook.presto.sql.analyzer.Analysis.addCoercion(Analysis.java:450)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitValues(StatementAnalyzer.java:1173)    <----- interesting, wrong branch
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitValues(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.Values.accept(Values.java:55)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:264)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:605)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:228)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTableSubquery(StatementAnalyzer.java:846)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTableSubquery(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.TableSubquery.accept(TableSubquery.java:53)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitAliasedRelation(StatementAnalyzer.java:788)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitAliasedRelation(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.AliasedRelation.accept(AliasedRelation.java:71)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeFrom(StatementAnalyzer.java:1689)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:860)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:264)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:605)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:242)
        at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:254)
        at com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:228)
        at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:72)
        at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:64)
        at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:305)
        at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:291)
        at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:247)
        at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
        at com.facebook.presto.execution.QueuedExecution$$Lambda$416.1022173988.run(Unknown Source:-1)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

Basically, the cause of the bug is assumptions that the only way an expression can have `row` type in `VALUES` is that `expression instanceof c.f.p.sql.tree.Row`, which is not true. Therefore, code shoud not depend on the Java type of `expression`. Instead, it should look at analysis.

**Comment by haozhun** [comment]
cc @martint. Can you check if my understanding for the expected behavior of this query is correct.

cc @rongrong for reporting this issue.

**Comment by martint** [comment]
If I recall correctly, this:

```
VALUES 
   (cast(row(1) as row(x bigint)))
```

is equivalent to:

```
VALUES
    ROW(cast(row(1) as row(x bigint)))
```

**Comment by haozhun** [comment]
It appears like SQL spec agrees with you, @martint. But I'm not confident about it.

```
<table value constructor> ::=
    VALUES <row value expression list>

<row value expression list> ::=
    <table row value expression> [ { <comma> <table row value expression> }... ]

<table row value expression> ::=
    <row value special case> | <row value constructor>

<row value special case> ::=
    <nonparenthesized value expression primary>

<row value constructor> ::=
    <common value expression> | <boolean value expression> | <explicit row value constructor>

<nonparenthesized value expression primary> ::=
    <unsigned value specification>
  | <column reference>
  | <set function specification>
  | <window function>
  | <nested window function>
  | <scalar subquery>
  | <case expression>
  | <cast specification>
  | <field reference>
  | <subtype treatment>
  | <method invocation>
  | <static method invocation>
  | <new specification>
  | <attribute or method reference>
  | <reference resolution>
  | <collection value constructor>
  | <array element reference>
  | <multiset element reference>
  | <next value expression>
  | <routine invocation>
  | <row pattern navigation operation>
  | <JSON value function>

<common value expression> ::=
    <numeric value expression>
  | <string value expression>
  | <datetime value expression>
  | <interval value expression>
  | <user-defined type value expression>
  | <reference value expression>
  | <collection value expression>

<explicit row value constructor> ::=
    <left paren> <row value constructor element> <comma> <row value constructor element list> <right paren>
  | ROW <left paren> <row value constructor element list> <right paren>
  | <row subquery>
```

* The declared type of a `<table row value expression>` is the declared type of the immediately contained `<row value special case>` or `<row value constructor>`.
* The declared type of a `<row value special case>` shall be a row type.
* If a `<row value constructor>` is a `<common value expression>` or a `<boolean value expression>` X, then the `<row value constructor>` is equivalent to ROW ( X )
* Let ERVC be an `<explicit row value constructor>`.  Case:
    1. If ERVC simply contains a `<row subquery>`, then the declared type of ERVC is the declared type of that `<row subquery>`.
    2. Otherwise, the declared type of ERVC is a row type described by a sequence of (`<field name>`, `<data type>`) pairs, corresponding in order to each `<row value constructor element>` X simply contained in ERVC. The `<data type>` is the declared type of X and the `<field name>` is implementation-dependent.

**Comment by martint** [comment]
Yeah, this covers what I was referring to:

> If a `<row value constructor>` is a `<common value expression>` or a `<boolean value expression>` X, then the `<row value constructor>` is equivalent to `ROW ( X )`

**Comment by haozhun** [comment]
Conclusion after discussion with @martint:

This query should produce a single column whose type is `row(x bigint)` because it falls into `<common value expression>`.
```
select * from (
  values 
    ( cast(row(1) as row(x bigint) )
) t(a);
```

This query should produce a single column whose type is `bigint` because it falls into `<nonparenthesized value expression primary>`.
```
select * from (
  values 
    cast(row(1) as row(x bigint))
) t(a);
```

**Comment by rongrong** [comment]
So we should fix this?

**Comment by losipiuk** [comment]
@haozhun parens do not match in your first example.
I guess it should be 
```
select * from (
  values 
    ( cast(row(1) as row(x bigint)))
) t(a);
```
But I did not follow the discussion
